tensorflow/compiler/xla/client/lib/matrix.cc

```diff
@@ -73,27 +75,52 @@ XlaOp GetMatrixDiagonal(XlaOp x, int k) {
     const int64 m = shape.dimensions(n_dims - 2);
     const int64 n = shape.dimensions(n_dims - 1);
 
-    auto mask = GetDiagonalMask(x, k);
-
-    // TPUs don't support S64 add reduction at the moment. But fortunately
-    // OR-reductions work just as well for integers.
-    XlaComputation reducer =
-        CreateScalarIdentityWithZeroComputation(shape.element_type(), builder);
-    // k == 0, we can save one slice op.
-    if (k == 0) {
-      return Reduce(Select(mask, x, Zeros(builder, shape)), ScalarLike(x, 0),
-                    reducer, {m >= n ? n_dims - 2 : n_dims - 1});
-    } else if (k > 0) {
-      auto result = Reduce(Select(mask, x, Zeros(builder, shape)),
-                           ScalarLike(x, 0), reducer, {n_dims - 2});
-      return SliceInMinorDims(result, {std::min<int64>(k, n)},
-                              {std::min(m + k, n)});
-    } else {
-      auto result = Reduce(Select(mask, x, Zeros(builder, shape)),
-                           ScalarLike(x, 0), reducer, {n_dims - 1});
-      return SliceInMinorDims(result, {std::min<int64>(-k, m)},
-                              {std::min(m, n - k)});
+    // The dimensions [0, axis) are batch dimensions.
```

Comment:
Can you add description like this to here and move the example below to right after this description?
The start_indices has a shape of {diag_len, 2}, and each pair of value in its dimension 1 represents the (row, col) of the diagonal. We set index_vector_dim to 1 and make start_index_map and collapsed_slice_dims contain the same two dimension indices. This makes sure that the (row, col) pairs in start_indices are propagated to the indices for the two collapsed dimensions in the operand indices through start_index_map.